### PR TITLE
Rework order targeting crazyness.

### LINKED ISSUE #7546: Unable to force fire on 3/4 cells on a bridge
If i force fire on the one cell that gives me an attack order icon the bridge gets destroyed but the units continue shooting after the bridge is destroyed. If I force fire at any of the other tree cells it will only show a move order icon and will only result in a move order unless there is a unit occupying that cell. 
The working cell:
116f1fa4-bceb-11e4-80f8-c55c6c46a78e
The none working cells:
22e8cce4-bceb-11e4-9d47-c9b897cd4ae2

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8908698-unable-to-force-fire-on-3-4-cells-on-a-bridge?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #10070: Cursors don't work correctly over bridges.
<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/28512890-cursors-don-t-work-correctly-over-bridges?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by pchote**
Reduces some duplication in UnitOrderGenerator and fixes #10070.  As a side effect, this adds the other half of the plumbing we need to fix #7546 (#9800 adds the first half), and gets us a small step closer to #8917.
#10070 was introduced by #9327 which gave them a finite selection size.  This meant that the ScreenMap query for actors under the cursor now returned the bridge actor instead of null, and meant that the move cursor and order queries checked against the actor's center position instead of the moused over cell.

This fix comes in two parts:
1. Adding a second pass to `UnitOrderGenerator.OrderForUnit` that checks for orders against the cell if there are no valid orders against the actor.
2. Making move orders invalid for any non-cell targets.

This fixes the cursor, and allows us to again order actors to any cell in the bridge footprint.  It also fixes the inconsistency where you got a move cursor when mousing over trees, but not over barrels.

This may have unexpected consequences, so it shouldn't be merged to prep:
- If there are any cases that relied on move orders against actors being valid when the moused-over cell was invalid but the actor's center position was valid (I don't think we have any code _this_ bogus... although it might affect crushing)
- If there are any other where we want to prevent orders against cells when there is an un-targetable actor on top.

### CODE REVIEW

**Comment by pchote** [comment]
Also fixes #9252.

File: OpenRA.Game/Orders/UnitOrderGenerator.cs [hunk 1]
```diff
@@ -112,27 +119,30 @@ static UnitOrderResult OrderForUnit(Actor self, Target target, MouseInput mi)
...
-					var modifiers = TargetModifiers.None;
-					if (mi.Modifiers.HasModifier(Modifiers.Ctrl))
-						modifiers |= TargetModifiers.ForceAttack;
-					if (mi.Modifiers.HasModifier(Modifiers.Shift))
-						modifiers |= TargetModifiers.ForceQueue;
-					if (mi.Modifiers.HasModifier(Modifiers.Alt))
-						modifiers |= TargetModifiers.ForceMove;
+			var orders = self.TraitsImplementing<IIssueOrder>()
+				.SelectMany(trait => trait.Orders.Select(x => new { Trait = trait, Order = x }))
+				.OrderByDescending(x => x.Order.OrderPriority);
```
Comment (RoosterDragon) [inline OpenRA.Game/Orders/UnitOrderGenerator.cs]: This may want a `ToList` to avoid being enumerated twice.

Comment (pchote) [inline OpenRA.Game/Orders/UnitOrderGenerator.cs]: I'm not convinced: in the common case we only enumerate this once, and in the worst case we enumerate it twice.

Comment (RoosterDragon) [inline OpenRA.Game/Orders/UnitOrderGenerator.cs]: I can agree with that.

**Comment by chrisforbes** [comment]
:+1: from reading only, not tested.

**Comment by RoosterDragon** [comment]
:+1: Code looks good. Bridges and badgers are indeed fixed and a cursory test reveals crushing is still possible.

**Comment by RoosterDragon** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/ca997d14e6d6f60fa0fab5fbbef74c2c272de30f) - `"Improved UI behaviour when issuing orders to units."`

**Pull request merged**